File: seqspec/seqspec_modify.py
"""Modify module for seqspec.

This module provides functionality to modify attributes of various elements in seqspec files.
"""

import json
from argparse import SUPPRESS, ArgumentParser, Namespace, RawTextHelpFormatter
from enum import Enum
from pathlib import Path
from typing import List

from seqspec.Assay import (
    Assay,
    AssayInput,
    LibKitInput,
    LibProtocolInput,
    SeqKitInput,
    SeqProtocolInput,
)
from seqspec.File import FileInput
from seqspec.Read import ReadInput
from seqspec.Region import RegionInput
from seqspec.utils import (
    load_assays,
    load_files,
    load_libkits,
    load_libprotocols,
    load_reads,
    load_regions,
    load_seqkits,
    load_seqprotocols,
    load_spec,
    write_pydantic_to_file_or_stdout,
)


def setup_modify_args(parser) -> ArgumentParser:
    """Create and configure the modify command subparser."""
    subparser = parser.add_parser(
        "modify",
        description="""
Modify attributes of various elements in a seqspec file using JSON objects.

Examples:
seqspec modify -m rna -o mod_spec.yaml -s read -k '[{"read_id": "rna_R1", "name": "renamed_rna_R1"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s region -k '[{"region_id": "rna_cell_bc", "name": "renamed_rna_cell_bc"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s file -k '[{"file_id": "R1.fastq.gz", "url": "./fastq/R1.fastq.gz"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s seqkit -k '[{"kit_id": "NovaSeq_kit", "name": "Updated NovaSeq Kit"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s seqprotocol -k '[{"protocol_id": "Illumina_NovaSeq", "name": "Updated Protocol"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s libkit -k '[{"kit_id": "Truseq_kit", "name": "Updated Truseq Kit"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s libprotocol -k '[{"protocol_id": "10x_protocol", "name": "Updated 10x Protocol"}]' spec.yaml
seqspec modify -m rna -o mod_spec.yaml -s assay -k '[{"assay_id": "my_assay", "name": "Updated Assay Name"}]' spec.yaml
---
""",
        help="Modify attributes of various elements in seqspec file",
        formatter_class=RawTextHelpFormatter,
    )
    subparser_required = subparser.add_argument_group("required arguments")
    subparser.add_argument("yaml", help="Sequencing specification yaml file", type=str)

    # TODO make this a potential path to a yaml file or json file
    subparser_required.add_argument(
        "-k",
        "--keys",
        metavar="KEYS",
        help="JSON array of objects to modify. Each object must include an id field (read_id, region_id, file_id, kit_id, protocol_id, or assay_id).",
        type=str,
        required=True,
    )

    subparser.add_argument(
        "-o",
        "--output",
        metavar="OUT",
        help="Path to output file",
        type=Path,
        default=None,
    )
    subparser_required.add_argument(
        "-i",
        metavar="IDs",
        help=SUPPRESS,
        type=str,
        default=None,
    )
    choices = [
        "read",
        "region",
        "file",
        "seqkit",
        "seqprotocol",
        "libkit",
        "libprotocol",
        "assay",
    ]
    subparser.add_argument(
        "-s",
        "--selector",
        metavar="SELECTOR",
        help=f"Selector for ID, [{', '.join(choices)}] (default: read)",
        type=str,
        default="read",
        choices=choices,
    )
    subparser_required.add_argument(
        "-m",
        "--modality",
        metavar="MODALITY",
        help="Modality of the assay",
        type=str,
        required=True,
    )

    return subparser


def validate_modify_args(parser: ArgumentParser, args: Namespace) -> None:
    """Validate the modify command arguments."""

    if not Path(args.yaml).exists():
        parser.error(f"Input file does not exist: {args.yaml}")

    if args.output and args.output.exists() and not args.output.is_file():
        parser.error(f"Output path exists but is not a file: {args.output}")


def run_modify(parser: ArgumentParser, args: Namespace) -> None:
    """Run the modify command."""
    validate_modify_args(parser, args)
    # todo enable updating id with args.id
    spec = load_spec(args.yaml)

    # Parse inline JSON string into a list of dicts; sub-functions will load objects
    raw_keys = json.loads(args.keys)
    if not isinstance(raw_keys, list):
        parser.error("--keys must be a JSON array of objects")

    keys = raw_keys

    spec = seqspec_modify(spec, args.modality, keys, args.selector)

    # Update spec
    spec.update_spec()

    write_pydantic_to_file_or_stdout(spec, args.output)


class Selector(str, Enum):
    READ = "read"
    REGION = "region"
    FILE = "file"
    SEQKIT = "seqkit"
    SEQPROTOCOL = "seqprotocol"
    LIBKIT = "libkit"
    LIBPROTOCOL = "libprotocol"
    ASSAY = "assay"


def seqspec_modify(
    spec: Assay,
    modality: str,
    keys: list[dict],
    selector: Selector,
) -> Assay:
    """Modify a loaded spec in-place according to selector-specific updates.

    This is the core implementation behind the ``seqspec modify`` CLI.
    It dispatches to a section-specific updater based on ``selector`` and
    applies partial updates to items identified by their stable IDs. Only
    fields present in each input object and not ``None`` are applied.

    Args:
        spec: Parsed ``Assay`` object to modify. Mutated in place and returned.
        modality: Target modality within the spec (for example, "rna", "atac").
        keys: List of JSON-like dictionaries describing updates. Each dict must
            include the identifying field for the chosen ``selector``:
            - selector == "read": use "read_id"
            - selector == "region": use "region_id"
            - selector == "file": use "file_id"
            - selector == "seqkit": use "kit_id"
            - selector == "seqprotocol": use "protocol_id"
            - selector == "libkit": use "kit_id"
            - selector == "libprotocol": use "protocol_id"
            - selector == "assay": use "assay_id" (must match ``spec.assay_id``)
        selector: Section of the spec to modify. See ``Selector`` enum.

    Returns:
        The same ``Assay`` instance with updates applied.

    Notes:
        - Read updates support the optional "files" field; when provided, the
          inputs are converted to ``File`` models before assignment.
        - Region updates are applied via ``Region.update_region_by_id`` within
          the library spec of the given modality.
        - Sequence/Library kits and protocols are updated field-by-field when
          the item with the matching ID is found; unknown IDs are ignored.
        - Unknown or missing identifiers in ``keys`` are skipped without error.

    Examples:
        Update a read name:
            >>> seqspec_modify(spec, "rna", [{"read_id": "rna_R1", "name": "R1_renamed"}], "read")

        Update a region sequence:
            >>> seqspec_modify(spec, "rna", [{"region_id": "rna_cell_bc", "sequence": "NNNN"}], "region")

        Update an assay description:
            >>> seqspec_modify(spec, "rna", [{"assay_id": spec.assay_id, "description": "New desc"}], "assay")
    """
    LOADERS = {
        Selector.READ: load_reads,
        Selector.REGION: load_regions,
        Selector.FILE: load_files,
        Selector.SEQKIT: load_seqkits,
        Selector.SEQPROTOCOL: load_seqprotocols,
        Selector.LIBKIT: load_libkits,
        Selector.LIBPROTOCOL: load_libprotocols,
        Selector.ASSAY: load_assays,
    }
    UPDATERS = {
        Selector.READ: seqspec_modify_read,
        Selector.REGION: seqspec_modify_region,
        Selector.FILE: seqspec_modify_files,
        Selector.SEQKIT: seqspec_modify_seqkit,
        Selector.SEQPROTOCOL: seqspec_modify_seqprotocol,
        Selector.LIBKIT: seqspec_modify_libkit,
        Selector.LIBPROTOCOL: seqspec_modify_libprotocol,
        Selector.ASSAY: seqspec_modify_assay,
    }

    selector_enum = selector if isinstance(selector, Selector) else Selector(selector)
    loader = LOADERS.get(selector_enum)
    updater = UPDATERS.get(selector_enum)
    if loader is None or updater is None:
        return spec

    loaded_inputs = loader(keys)
    mod_spec: Assay = updater(spec, modality, loaded_inputs)  # type: ignore[arg-type]
    return mod_spec


def seqspec_modify_read(
    spec: Assay, modality: str, new_reads: List[ReadInput]
) -> Assay:
    """Modify read properties in spec using ``ReadInput`` objects."""
    reads = spec.get_seqspec(modality)
    for patch in new_reads:
        read_id = getattr(patch, "read_id", None)
        if not read_id:
            continue
        for read in reads:
            if read.read_id == read_id:
                files_value = None
                if "files" in patch.model_fields_set and patch.files is not None:
                    files_value = [f.to_file() for f in patch.files]
                read.update_read_by_id(
                    read_id=patch.read_id,
                    name=patch.name,
                    modality=patch.modality,
                    primer_id=patch.primer_id,
                    min_len=patch.min_len,
                    max_len=patch.max_len,
                    strand=patch.strand,
                    files=files_value,
                )
                break
    return spec


def seqspec_modify_region(
    spec: Assay, modality: str, new_regions: List[RegionInput]
) -> Assay:
    """Modify region properties in spec using ``RegionInput`` objects."""
    libspec = spec.get_libspec(modality)
    for patch in new_regions:
        region_id = getattr(patch, "region_id", None)
        if not region_id:
            continue
        libspec.update_region_by_id(
            target_region_id=region_id,
            region_id=patch.region_id,
            region_type=patch.region_type,
            name=patch.name,
            sequence_type=patch.sequence_type,
            sequence=patch.sequence,
            min_len=patch.min_len,
            max_len=patch.max_len,
        )
    return spec.model_copy()


def seqspec_modify_files(
    spec: Assay, modality: str, new_files: List[FileInput]
) -> Assay:
    """Modify file properties in spec using ``FileInput`` objects."""
    reads = spec.get_seqspec(modality)
    for patch in new_files:
        file_id = getattr(patch, "file_id", None)
        if not file_id:
            continue
        for read in reads:
            for file in read.files:
                if file.file_id == file_id:
                    for field in patch.model_fields_set:
                        value = getattr(patch, field)
                        if value is not None:
                            setattr(file, field, value)
                    break
    return spec


def seqspec_modify_seqkit(
    spec: Assay, modality: str, new_seqkits: List[SeqKitInput]
) -> Assay:
    """Modify sequence kit properties in spec using ``SeqKitInput`` objects."""
    if isinstance(spec.sequence_kit, list):
        for patch in new_seqkits:
            kit_id = getattr(patch, "kit_id", None)
            if not kit_id:
                continue
            for seqkit in spec.sequence_kit:
                if seqkit.kit_id == kit_id:
                    for field in patch.model_fields_set:
                        value = getattr(patch, field)
                        if value is not None and hasattr(seqkit, field):
                            setattr(seqkit, field, value)
                    break
    return spec


def seqspec_modify_seqprotocol(
    spec: Assay,
    modality: str,
    new_seqprotocols: List[SeqProtocolInput],
) -> Assay:
    """Modify sequence protocol properties using ``SeqProtocolInput`` objects."""
    if isinstance(spec.sequence_protocol, list):
        for patch in new_seqprotocols:
            protocol_id = getattr(patch, "protocol_id", None)
            if not protocol_id:
                continue
            for seqprotocol in spec.sequence_protocol:
                if seqprotocol.protocol_id == protocol_id:
                    for field in patch.model_fields_set:
                        value = getattr(patch, field)
                        if value is not None and hasattr(seqprotocol, field):
                            setattr(seqprotocol, field, value)
                    break
    return spec


def seqspec_modify_libkit(
    spec: Assay, modality: str, new_libkits: List[LibKitInput]
) -> Assay:
    """Modify library kit properties in spec using ``LibKitInput`` objects."""
    if isinstance(spec.library_kit, list):
        for patch in new_libkits:
            kit_id = getattr(patch, "kit_id", None)
            if not kit_id:
                continue
            for libkit in spec.library_kit:
                if libkit.kit_id == kit_id:
                    for field in patch.model_fields_set:
                        value = getattr(patch, field)
                        if value is not None and hasattr(libkit, field):
                            setattr(libkit, field, value)
                    break
    return spec


def seqspec_modify_libprotocol(
    spec: Assay,
    modality: str,
    new_libprotocols: List[LibProtocolInput],
) -> Assay:
    """Modify library protocol properties using ``LibProtocolInput`` objects."""
    if isinstance(spec.library_protocol, list):
        for patch in new_libprotocols:
            protocol_id = getattr(patch, "protocol_id", None)
            if not protocol_id:
                continue
            for libprotocol in spec.library_protocol:
                if libprotocol.protocol_id == protocol_id:
                    for field in patch.model_fields_set:
                        value = getattr(patch, field)
                        if value is not None and hasattr(libprotocol, field):
                            setattr(libprotocol, field, value)
                    break
    return spec


def seqspec_modify_assay(
    spec: Assay, modality: str, new_assay_data: List[AssayInput]
) -> Assay:
    """Modify assay properties using ``AssayInput`` objects."""
    for patch in new_assay_data:
        assay_id = getattr(patch, "assay_id", None)
        if not assay_id or assay_id != spec.assay_id:
            continue
        for field in patch.model_fields_set:
            value = getattr(patch, field)
            if value is not None and hasattr(spec, field):
                setattr(spec, field, value)
    return spec

File: src/seqspec_modify.rs
use crate::models::assay::Assay;
use crate::models::file::File;
use crate::models::region_type::RegionTypeValue;
use crate::utils;
use clap::Args;
use serde_json::Value;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Args)]
pub struct ModifyArgs {
    #[clap(help = "Sequencing specification yaml file", required = true)]
    yaml: PathBuf,

    #[clap(
        short,
        long,
        help = "Modality of the assay",
        value_name = "MODALITY",
        required = true
    )]
    modality: String,

    #[clap(
        short,
        long,
        help = "JSON array of objects to modify",
        value_name = "KEYS",
        required = true
    )]
    keys: String,

    #[clap(short = 'i', hide = true, value_name = "IDs")]
    legacy_ids: Option<String>,

    #[clap(
        short,
        long,
        help = "Selector",
        value_name = "SELECTOR",
        default_value = "read",
        value_parser = ["read","region","file","seqkit","seqprotocol","libkit","libprotocol","assay"]
    )]
    selector: String,

    #[clap(short, long, help = "Path to output file", value_name = "OUT")]
    output: Option<PathBuf>,
}

pub fn run_modify(args: &ModifyArgs) {
    validate_modify_args(args);
    let mut spec = utils::load_spec(&args.yaml);

    let keys: Vec<Value> = serde_json::from_str(&args.keys).expect("--keys must be a JSON array");
    let selector = args.selector.as_str();
    spec = seqspec_modify(spec, &args.modality, keys, selector);

    spec.update_spec();

    let yaml = spec.to_bytes().unwrap();
    if let Some(out) = &args.output {
        let mut f = fs::File::create(out).unwrap();
        f.write_all(&yaml).unwrap();
    } else {
        println!("{}", String::from_utf8_lossy(&yaml));
    }
}

fn validate_modify_args(args: &ModifyArgs) {
    if !args.yaml.exists() {
        eprintln!("Please use `seqspec modify -h` for help.");
        std::process::exit(1);
    }
    if let Some(out) = &args.output {
        if out.exists() && !out.is_file() {
            eprintln!("Output path exists but is not a file: {}", out.display());
            std::process::exit(1);
        }
    }
}

pub fn seqspec_modify(mut spec: Assay, modality: &str, keys: Vec<Value>, selector: &str) -> Assay {
    match selector {
        "read" => modify_reads(&mut spec, modality, &keys),
        "region" => modify_regions(&mut spec, modality, &keys),
        "file" => modify_files(&mut spec, modality, &keys),
        "seqkit" => modify_seqkits(&mut spec, &keys),
        "seqprotocol" => modify_seqprotocols(&mut spec, &keys),
        "libkit" => modify_libkits(&mut spec, &keys),
        "libprotocol" => modify_libprotocols(&mut spec, &keys),
        "assay" => modify_assay(&mut spec, &keys),
        _ => (),
    }
    spec
}

fn vstr(v: &Value, key: &str) -> Option<String> {
    v.get(key).and_then(|x| x.as_str().map(|s| s.to_string()))
}

fn vregion_type(v: &Value, key: &str) -> Option<RegionTypeValue> {
    v.get(key)
        .and_then(|value| serde_json::from_value(value.clone()).ok())
}

fn vi64(v: &Value, key: &str) -> Option<i64> {
    v.get(key).and_then(|x| x.as_i64())
}

fn modify_reads(spec: &mut Assay, modality: &str, keys: &Vec<Value>) {
    for patch in keys {
        let Some(read_id) = vstr(patch, "read_id") else {
            continue;
        };
        if let Some(rd) = spec
            .sequence_spec
            .iter_mut()
            .find(|r| r.modality == modality && r.read_id == read_id)
        {
            // files optional
            let files_opt: Option<Vec<File>> = patch.get("files").and_then(|arr| {
                arr.as_array().map(|items| {
                    items
                        .iter()
                        .filter_map(|it| {
                            Some(File::new(
                                it.get("file_id")?.as_str()?.to_string(),
                                it.get("filename")
                                    .and_then(|x| x.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                                it.get("filetype")
                                    .and_then(|x| x.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                                it.get("filesize").and_then(|x| x.as_i64()).unwrap_or(0),
                                it.get("url")
                                    .and_then(|x| x.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                                it.get("urltype")
                                    .and_then(|x| x.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                                it.get("md5")
                                    .and_then(|x| x.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                            ))
                        })
                        .collect()
                })
            });

            rd.update_read_by_id(
                vstr(patch, "read_id"),
                vstr(patch, "name"),
                vstr(patch, "modality"),
                vstr(patch, "primer_id"),
                vi64(patch, "min_len"),
                vi64(patch, "max_len"),
                vstr(patch, "strand"),
                files_opt,
            );
        }
    }
}

fn modify_regions(spec: &mut Assay, modality: &str, keys: &Vec<Value>) {
    // find index for modality to get mutable region tree
    if let Some(idx) = spec.modalities.iter().position(|m| m == modality) {
        if let Some(target) = spec.library_spec.get_mut(idx) {
            for patch in keys {
                let Some(target_region_id) = vstr(patch, "region_id") else {
                    continue;
                };
                target.update_region_by_id(
                    target_region_id,
                    vstr(patch, "region_id"),
                    vregion_type(patch, "region_type"),
                    vstr(patch, "name"),
                    vstr(patch, "sequence_type"),
                    vstr(patch, "sequence"),
                    vi64(patch, "min_len"),
                    vi64(patch, "max_len"),
                );
            }
        }
    }
}

fn modify_files(spec: &mut Assay, modality: &str, keys: &Vec<Value>) {
    for patch in keys {
        let Some(file_id) = vstr(patch, "file_id") else {
            continue;
        };
        for r in spec
            .sequence_spec
            .iter_mut()
            .filter(|r| r.modality == modality)
        {
            for f in &mut r.files {
                if f.file_id == file_id {
                    if let Some(v) = vstr(patch, "filename") {
                        f.filename = v;
                    }
                    if let Some(v) = vstr(patch, "filetype") {
                        f.filetype = v;
                    }
                    if let Some(v) = vi64(patch, "filesize") {
                        f.filesize = v;
                    }
                    if let Some(v) = vstr(patch, "url") {
                        f.url = v;
                    }
                    if let Some(v) = vstr(patch, "urltype") {
                        f.urltype = v;
                    }
                    if let Some(v) = vstr(patch, "md5") {
                        f.md5 = v;
                    }
                }
            }
        }
    }
}

fn modify_seqkits(spec: &mut Assay, keys: &Vec<Value>) {
    if let Some(kits) = spec.sequence_kit.as_mut() {
        for patch in keys {
            let Some(kit_id) = vstr(patch, "kit_id") else {
                continue;
            };
            if let Some(k) = kits.iter_mut().find(|k| k.kit_id == kit_id) {
                if let Some(v) = vstr(patch, "name") {
                    k.name = Some(v);
                }
                if let Some(v) = vstr(patch, "modality") {
                    k.modality = v;
                }
            }
        }
    }
}

fn modify_seqprotocols(spec: &mut Assay, keys: &Vec<Value>) {
    if let Some(protocols) = spec.sequence_protocol.as_mut() {
        for patch in keys {
            let Some(protocol_id) = vstr(patch, "protocol_id") else {
                continue;
            };
            if let Some(p) = protocols.iter_mut().find(|p| p.protocol_id == protocol_id) {
                if let Some(v) = vstr(patch, "name") {
                    p.name = v;
                }
                if let Some(v) = vstr(patch, "modality") {
                    p.modality = v;
                }
            }
        }
    }
}

fn modify_libkits(spec: &mut Assay, keys: &Vec<Value>) {
    if let Some(kits) = spec.library_kit.as_mut() {
        for patch in keys {
            let Some(kit_id) = vstr(patch, "kit_id") else {
                continue;
            };
            if let Some(k) = kits.iter_mut().find(|k| k.kit_id == kit_id) {
                if let Some(v) = vstr(patch, "name") {
                    k.name = Some(v);
                }
                if let Some(v) = vstr(patch, "modality") {
                    k.modality = v;
                }
            }
        }
    }
}

fn modify_libprotocols(spec: &mut Assay, keys: &Vec<Value>) {
    if let Some(protocols) = spec.library_protocol.as_mut() {
        for patch in keys {
            let Some(protocol_id) = vstr(patch, "protocol_id") else {
                continue;
            };
            if let Some(p) = protocols.iter_mut().find(|p| p.protocol_id == protocol_id) {
                if let Some(v) = vstr(patch, "name") {
                    p.name = v;
                }
                if let Some(v) = vstr(patch, "modality") {
                    p.modality = v;
                }
            }
        }
    }
}

fn modify_assay(spec: &mut Assay, keys: &Vec<Value>) {
    for patch in keys {
        let Some(assay_id) = vstr(patch, "assay_id") else {
            continue;
        };
        if assay_id != spec.assay_id {
            continue;
        }
        if let Some(v) = vstr(patch, "name") {
            spec.name = v;
        }
        if let Some(v) = vstr(patch, "doi") {
            spec.doi = v;
        }
        if let Some(v) = vstr(patch, "date") {
            spec.date = v;
        }
        if let Some(v) = vstr(patch, "description") {
            spec.description = v;
        }
        if let Some(v) = vstr(patch, "lib_struct") {
            spec.lib_struct = v;
        }
        if let Some(v) = vstr(patch, "assay_id") {
            spec.assay_id = v;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::load_spec;
    use serde_json::json;

    fn dogma_spec() -> Assay {
        load_spec(&PathBuf::from("tests/fixtures/spec.yaml"))
    }

    #[test]
    fn test_modify_read_name() {
        let spec = dogma_spec();
        let rna_reads = spec.get_seqspec("rna");
        let read_id = rna_reads[0].read_id.clone();
        let keys = vec![json!({"read_id": read_id, "name": "Updated Name"})];
        let modified = seqspec_modify(spec, "rna", keys, "read");
        let read = modified.get_read(&read_id).unwrap();
        assert_eq!(read.name, "Updated Name");
    }

    #[test]
    fn test_modify_region_name() {
        let spec = dogma_spec();
        let lib = spec.get_libspec("rna").unwrap();
        let leaves = lib.get_leaves();
        let target = &leaves[0];
        let keys = vec![json!({"region_id": target.region_id, "name": "New Name"})];
        let modified = seqspec_modify(spec, "rna", keys, "region");
        let lib = modified.get_libspec("rna").unwrap();
        let found = lib.get_region_by_id(&target.region_id);
        assert!(!found.is_empty());
        assert_eq!(found[0].name, "New Name");
    }

    #[test]
    fn test_modify_region_accepts_region_type_list() {
        let spec = dogma_spec();
        let keys = vec![json!({"region_id": "rna_cell_bc", "region_type": ["RGN:partition:cell"]})];
        let modified = seqspec_modify(spec, "rna", keys, "region");
        let region = modified
            .get_libspec("rna")
            .unwrap()
            .get_region_by_id("rna_cell_bc")[0]
            .clone();
        assert!(region.region_type.matches("barcode"));
        assert!(region.region_type.matches("RGN:partition:cell"));
    }

    #[test]
    fn test_modify_file_url() {
        let spec = dogma_spec();
        let rna_reads = spec.get_seqspec("rna");
        // Find a read with files
        let read_with_files = rna_reads.iter().find(|r| !r.files.is_empty());
        if let Some(rd) = read_with_files {
            let file_id = rd.files[0].file_id.clone();
            let keys = vec![json!({"file_id": file_id, "url": "http://new.url/file.fq.gz"})];
            let modified = seqspec_modify(spec, "rna", keys, "file");
            let updated_read = modified.get_read(&rd.read_id).unwrap();
            let f = updated_read
                .files
                .iter()
                .find(|f| f.file_id == file_id)
                .unwrap();
            assert_eq!(f.url, "http://new.url/file.fq.gz");
        }
    }

    #[test]
    fn test_modify_assay_fields() {
        let spec = dogma_spec();
        let assay_id = spec.assay_id.clone();
        let keys = vec![
            json!({"assay_id": assay_id, "name": "New Assay Name", "description": "Updated desc"}),
        ];
        let modified = seqspec_modify(spec, "rna", keys, "assay");
        assert_eq!(modified.name, "New Assay Name");
        assert_eq!(modified.description, "Updated desc");
    }

    #[test]
    fn test_modify_unknown_selector() {
        let spec = dogma_spec();
        let keys = vec![json!({"id": "x"})];
        let modified = seqspec_modify(spec.clone(), "rna", keys, "unknown");
        assert_eq!(modified.assay_id, spec.assay_id);
    }

    #[test]
    fn test_modify_read_preserves_global_read_order() {
        let spec = dogma_spec();
        let original_ids: Vec<String> = spec
            .sequence_spec
            .iter()
            .map(|read| read.read_id.clone())
            .collect();
        let keys = vec![json!({"read_id": "rna_R1", "name": "Updated Name"})];
        let modified = seqspec_modify(spec, "rna", keys, "read");
        let modified_ids: Vec<String> = modified
            .sequence_spec
            .iter()
            .map(|read| read.read_id.clone())
            .collect();
        assert_eq!(modified_ids, original_ids);
    }
}
